### PR TITLE
Implement device-side RNG

### PR DESCRIPTION
**Pull request by utkarsh530**
```
using AMDGPU
using Random
function kernel()
    Random.seed!(123)
    return nothing
end
@roc kernel()
```

```
julia> @roc kernel()
ERROR: GPU compilation of kernel #kernel() failed
KernelError: kernel returns a value of type `Union{}`

Make sure your kernel function ends in `return`, `return nothing` or `nothing`.
If the returned value is of type `Union{}`, your Julia code probably throws an exception.
Inspect the code with `@device_code_warntype` for more details.

Stacktrace:
 [1] check_method(job::GPUCompiler.CompilerJob)
   @ GPUCompiler ~/.julia/packages/GPUCompiler/kb6yJ/src/validation.jl:41
 [2] macro expansion
   @ ~/.julia/packages/TimerOutputs/LHjFw/src/TimerOutput.jl:253 [inlined]
 [3] macro expansion
   @ ~/.julia/packages/GPUCompiler/kb6yJ/src/driver.jl:152 [inlined]
 [4] emit_julia(job::GPUCompiler.CompilerJob)
   @ GPUCompiler ~/.julia/packages/GPUCompiler/kb6yJ/src/utils.jl:83
 [5] rocfunction_compile(job::GPUCompiler.CompilerJob)
   @ AMDGPU.Compiler ~/.julia/dev/AMDGPU/src/compiler/codegen.jl:181
 [6] cached_compilation(cache::Dict{UInt64, Any}, job::GPUCompiler.CompilerJob, compiler::typeof(AMDGPU.Compiler.rocfunction_compile), linker::typeof(AMDGPU.Compiler.rocfunction_link))
   @ GPUCompiler ~/.julia/packages/GPUCompiler/kb6yJ/src/cache.jl:90
 [7] rocfunction(f::typeof(kernel), tt::Type; name::Nothing, device::ROCDevice, global_hooks::NamedTuple{(), Tuple{}})
   @ AMDGPU.Compiler ~/.julia/dev/AMDGPU/src/compiler/codegen.jl:165
 [8] rocfunction(f::typeof(kernel), tt::Type)
   @ AMDGPU.Compiler ~/.julia/dev/AMDGPU/src/compiler/codegen.jl:154
 [9] top-level scope
   @ ~/.julia/dev/AMDGPU/src/highlevel.jl:430

```

### CODE REVIEW

**Comment by utkarsh530** [comment]
Intended to Fix #378

**Comment by williamfgc** [comment]
@utkarsh530 @jpsamaroo thanks for the PR! Is there anything on my side to help merge this?

File: src/device/random.jl [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        $("""@__zeroinit_global_random_counters = external addrspace($(AS.Local)) global [32 x i32], align 32
+             define i8 addrspace($(AS.Local))* @entry() #0 {
+                 %ptr = getelementptr inbounds [32 x i32], [32 x i32] addrspace($(AS.Local))* @__zeroinit_global_random_counters, i64 0, i64 0
+                 %untyped_ptr = bitcast i32 addrspace($(AS.Local))* %ptr to i8 addrspace($(AS.Local))*
+                 ret i8 addrspace($(AS.Local))* %untyped_ptr
+             }
+             attributes #0 = { alwaysinline }
+          """, "entry"), LLVMPtr{UInt32, AS.Local}, Tuple{})
+    ROCDeviceArray{UInt32,1,AS.Local}((32,), ptr)
+end
```
Comment (vchuravy) [inline src/device/random.jl:38]: Can we use LLVM.jl to build these? E.g. to be robust against IR changes

Comment (jpsamaroo) [reply src/device/random.jl]: Implemented in the latest commit

File: src/device/random.jl [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    end
+end
+
+# default to 7 rounds; enough to pass SmallCrush
+@inline Philox2x32() = Philox2x32{7}()
+
+@inline function Base.getproperty(rng::Philox2x32, field::Symbol)
+    threadId = workitemIdx().x + (workitemIdx().y - Int32(1)) * workgroupDim().x +
+                                 (workitemIdx().z - Int32(1)) * workgroupDim().x * workgroupDim().y
+    warpId = (threadId - Int32(1)) >> 0x6 + Int32(1)  # fld1 by 64
```
[the remark this answers is not in the document]
Comment (pxl-th) [reply src/device/random.jl:87]: Also, Navi 2 & 3 have wavefront of size 32.
So if we really want to be correct, we probably should rely on `Device.wavefrontsize()`.

File: src/device/random.jl [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                                 (workgroupIdx().z - Int32(1)) * gridGroupDim().x * gridGroupDim().y
+        globalId = threadId + (blockId - Int32(1)) * (workgroupDim().x * workgroupDim().y * workgroupDim().z)
+        globalId%UInt32
+    end::UInt32
+end
+
+@inline function Base.setproperty!(rng::Philox2x32, field::Symbol, x)
+    threadId = workitemIdx().x + (workitemIdx().y - Int32(1)) * workgroupDim().x +
+                                 (workitemIdx().z - Int32(1)) * workgroupDim().x * workgroupDim().y
+    warpId = (threadId - Int32(1)) >> 0x6 + Int32(1)  # fld1 by 64
```
Comment (pxl-th) [inline src/device/random.jl:106]:
```suggestion
    warpId = (threadId - Int32(1)) >> 0x5 + Int32(1)  # fld1 by 32
```

**Pull request merged**